pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java

```diff
@@ -224,11 +226,13 @@
         ObjectMapper objectMapper = ObjectMapperFactory.getThreadLocal();
         String configContent = objectMapper.writeValueAsString(goInstanceConfig);
 
-        // Nit: at present, the implementation of go function depends on pulsar-client-go,
-        // pulsar-client-go uses cgo, so the currently uploaded executable doesn't support cross-compilation.
         args.add(originalCodeFileName);
         args.add("-instance-conf");
-        args.add("'" + configContent + "'");
+        if (k8sRuntime) {
```

Comment:
Why cannot we use a similar approach as for the Java and Python runtimes (`JsonFormat.printer`)?

https://github.com/apache/pulsar/blob/a045d926c9620a3744a4bc755bf6f2a65209e6b6/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java#L349


You're fixing the issue for non-k8s runtimes, but wouldn't the problem with json parsing still occur in k8s runtime?

Reply:
Go Runtime is different from Python and Java Functions runtime, this is mainly the language itself. Go Runtime separately encapsulates the GoInstanceConfig object to process the data transmitted from FunctionDetails, so we can’t use the methods carried by protobuf itself to parse . In Go Runtime, we use `-instance-conf` to receive and process all parameter information.

Reply:
But we could use a similar quote escaping mechanism with JsonPrinter, couldn't we?
My concern that your solution might still produce an incorrectly escaped string if k8sruntime == true.
It'd be great to have some integration tests that verify a function that has user-config can be successfully started in K8S runtime.

Reply:
> But we could use a similar quote escaping mechanism with JsonPrinter, couldn't we?

This library is provided by k8s, which is not suitable for Go Runtime here, right?

Reply:
Again, In go runtime, we have no way to directly parse the parameters passed in from the command line. This is limited by go instance. We converted all the configuration to `GoInstanceConfig` for processing. The reason for this problem is because in the runtime of the process , Due to the introduction of single quotation marks in config content, the problem of nesting of single quotation marks may occur when creating functions.

Reply:
I realized JsonPrinter accepts protobuf's MessageOrBuilder, fair enough.

Still, we have to be sure that a string like `--user-config='{"foo": "bar"}'` is propagated correctly when wrapped into single quotes.

Here's a simple example:

```bash
# Current approach: single quotes in the json get removed after wrapping the clause with additional single quotes. 
➜ echo '--user-config='{"foo": "bar"}''
➜ --user-config={foo: bar}


# Possible option: prefix with $ and use escaping \'
➜ echo $'--user-config=\'{"foo": "bar"}\''
➜ --user-config='{"foo": "bar"}'
```

What do you think?

Reply:
> Still, we have to be sure that a string like --user-config='{"foo": "bar"}' is propagated correctly when wrapped into single quotes.

Why? As far as I understand it, `--user-config` must be JSON. Strings that begin with single quotes are invalid JSON, full stop. If you specify this on the command line, your shell will take care of this, and if you put it in that value on purpose, then your function will fail and it is your fault.

Reply:
Having looked into it a bit more, I see now that the k8s runtime shells out to issue some cmod hacks and run `exec` on the go binary itself. While this seems suboptimal, I at least understand what's going on a bit better now. I was talking with @addisonj a bit about this and would it make more sense to build the arguments and then only later post process them to add quotes if the runtime is k8s?

Reply:
Thanks @vaihtovirta @flowchartsman work for this, If we have a better solution, we can send an issue to track this issue.

Reply:
@flowchartsman I see, I realized `-instance-conf` takes a json string as an argument, so my previous message doesn't make sense.